src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -729,153 +728,291 @@ func (s *service) FetchTagged(tctx thrift.Context, req *rpc.FetchTaggedRequest)
-	// but don't serialize yet so that all block reader requests can
-	// be issued at once before waiting for their results.
-	var encodedDataResults [][][]xio.BlockReader
-	if fetchData {
-		encodedDataResults = make([][][]xio.BlockReader, results.Size())
-	}
-	if err := s.fetchReadEncoded(ctx, db, response, results, ns, nsIDBytes, callStart, opts, fetchData, encodedDataResults); err != nil {
-		s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
-		return nil, err
-	}
+	return newFetchTaggedResultsIter(fetchTaggedResultsIterOpts{
+		queryResult: queryResult,
+		queryOpts:   opts,
+		fetchData:   fetchData,
+		db:          db,
+		docReader:   docs.NewEncodedDocumentReader(),
+		nsID:        ns,
+		tagEncoder:  tagEncoder,
+		iOpts:       s.opts.InstrumentOptions(),
+	}), nil
+}
 
-	// Step 2: If fetching data read the results of the asynchronous block readers.
-	if fetchData {
-		if err := s.fetchReadResults(ctx, response, ns, encodedDataResults); err != nil {
-			s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
-			return nil, err
-		}
-	}
+// BaseIter has common iterator methods.
+type BaseIter interface {
+	// Next advances to the next element. Returns false if all elements have already been processed. If an error occurs
+	// it can be retrieved with Err().
+	Next(ctx context.Context) bool
 
-	s.metrics.fetchTagged.ReportSuccess(s.nowFn().Sub(callStart))
-	return response, nil
+	// Err returns a non-nil error if an error occurred when calling Next().
+	Err() error
 }
 
-func (s *service) fetchReadEncoded(ctx context.Context,
-	db storage.Database,
-	response *rpc.FetchTaggedResult_,
-	results index.QueryResults,
-	nsID ident.ID,
-	nsIDBytes []byte,
-	callStart time.Time,
-	opts index.QueryOptions,
-	fetchData bool,
-	encodedDataResults [][][]xio.BlockReader,
-) error {
-	ctx, sp, sampled := ctx.StartSampledTraceSpan(tracepoint.FetchReadEncoded)
-	if sampled {
-		sp.LogFields(
-			opentracinglog.String("id", nsID.String()),
-			xopentracing.Time("callStart", callStart),
-		)
+// FetchTaggedResultsIter iterates over the results from FetchTagged
+// The iterator is not thread safe and must only be accessed from a single goroutine.
+type FetchTaggedResultsIter interface {
+	BaseIter
+
+	// NumIDs returns the total number of series IDs in the result.
+	NumIDs() int
+
+	// Exhaustive returns true if NumIDs is all IDs that the query could have returned.
+	Exhaustive() bool
+
+	// Namespace is the namespace.
+	Namespace() ident.ID
+
+	// Current returns the current IDResult. The result is only valid if Err() is nil.
+	Current() IDResult
+}
+
+// SegmentsIter iterates over the Segments for an IDResult.
+type SegmentsIter interface {
+	BaseIter
+	// Current returns the current Segments. The result is only valid if Err() is nil.
+	Current() *rpc.Segments
+}
+
+type fetchTaggedResultsIter struct {
+	queryResults   map[index.ResultsMapHash]index.ResultsMapEntry
```

Comment:
Q: Why not take the `*index.ResultsMap` itself here? Not that this isn't wrong it just breaks the abstraction a little bit of our map type wrapping the underlying map itself (in case we ever wanted to change the way the `.Iter()` method worked for the map type).